테스트 작성하기

alex-gibeom opened the pull request:
안녕하세요 리뷰어님 :)

드디어 테스트를 시작하네요. 너무 재밌을 것 같습니다..!!
테스트는 한번도 경험해보지 못해서 많이 부족합니다 😂
이번주도 열심히 공부하며 따라가보겠습니다 😃

이번 한 주 잘 부탁드립니다!! 🙇‍♂️🙇‍♂️

alex-gibeom committed:
feat: Task 모델 테스트 완료

alex-gibeom committed:
feat: TaskService 테스트 완료

alex-gibeom committed:
merge: [Day-1] 단일 테스트

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,118 @@
+        assertThat(task.getTitle()).isEqualTo(TEST_TITLE);
+    }
+
+    @Test
+    void getTasks() {
```
각 테스트별로 `@DisplayName`을 사용해서 테스트의 목적과 기대하는 결과에 대해 설명을 적어 보도록 합시다.

alex-gibeom replied:
(7573917) 해당 수정에서 `@DisplayName` 과 `@DisplayNameGeneration`을 통해 테스트의 이름을 작성해보았습니다!

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,118 @@
+    private final String TEST_TITLE = "테스트는 재밌군요!";
+    private final String POSTFIX_TITLE = " 그치만 매우 생소하군요!";
+
+    @BeforeEach
```
`@BeforeEach` 사용법을 이미 알고 계시는군요!

alex-gibeom replied:
강의에서 나온 것을 토대로 공부하며 사용해보았습니다!

추가로 [공식 문서](https://junit.org/junit5/docs/current/user-guide/#writing-tests-display-names)를 통해 학습한 것으로는 `@BeforeEach`는 아래와 같은 어노테이션이 달린 메서드보다 먼저 실행된다고 하네요 :)
- `@Test`
- `@RepeatedTest`
- `@ParameterizedTest`
- `@TestFactory`

<br>

### 질문
`@BeforeEach`를 통해 Fixture 데이터를 생성해 보았습니다!

강의에서는 Fixture가 **믿고 사용할 수 있는 데이터**를 미리 만들어 놓고, 테스트에 사용하는 것으로 들었습니다!
따라서 아래와 같이 baseTask를 생성하는 메서드에도 테스트 검증(`assertThat`)을 진행하였는데, 보통 어떤식으로 하는지 궁금합니다!
```java
    @BeforeEach
    void setUp() {
        taskService = new TaskService();

        createBaseTask();
    }

    /**
     * 각 테스트를 위한 fixture 데이터 생성
     */
    void createBaseTask() {
        Task source = new Task();
        source.setTitle(TEST_TITLE);
        Task task = taskService.createTask(source);

        assertThat(task.getId()).isEqualTo(TEST_ID);
        assertThat(task.getTitle()).isEqualTo(TEST_TITLE);
    }
```

johngrib replied:
프로젝트의 상황과 목적에 따라 다양한 방법을 씁니다. @giibeom 님이 쓴 것 같은 방법도 자주 쓰입니다. 하지만 자주 쓰인다고 좋은 방법인 것은 아닙니다. 테스트를 준비하는 곳에서는 준비하는 작업만 하고, 테스트를 검증하는 곳에서만 검증할 수 있도록 합시다. 역할을 섞는 것은 좋지 않습니다.

alex-gibeom replied:
뭔가 `@Test`가 아닌데 `assertThat`을 사용하는 것이 부자연스럽다고 문득 생각했는데 자주 쓰이긴 하는군요 :)

역할을 섞지 않는 것이 좋을 것 같다는 생각이 들면서,
한편으로는 fixture 데이터가 신뢰성이  있어야 이를 사용하는  테스트들도 신뢰성이 있다고 생각합니다.

"**아래에서 `@Test`를 통해 `createTask`의 테스트를 진행함으로써, `setUp()`에서의 fixture 데이터를 넣는 행위도 신뢰할 수 있다**" 라는 관점으로 보면 되는 것일까요?

johngrib replied:
그런 것들은 팀의 개발철학이나 목표에 따라 조금씩 달라지죠. 그래서 어떤 곳에서는 이 방법을 쓰고 어떤 곳에서는 금지할 것입니다. 일단 코드숨에서는 사용하지 않는 방향으로 컨벤션이 정해졌다고 생각하고 접근해 보도록 합시다. 우리는 앞으로 꽤나 엄격하게 테스트 공간을 구조적으로 분리하며 작업하는 것을 연습할 것이기 때문입니다.

alex-gibeom replied:
알겠습니다 :)
답변 감사드립니다 🙇‍♂️

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,118 @@
+
+    @Test
+    void getTaskWithValidId() {
+        Task task = taskService.getTask(1L);
+
+        // fixture 데이터의 id가 정상인지 확인
+        assertThat(task.getId()).isEqualTo(1L);
```
두 `1L`이 직접 언급된 것이 아쉽습니다. 지금은 몇 줄짜리 짧은 메소드라 이런 방식이 별로 문제가 안 됩니다. 하지만 10줄이 넘어가고 이런 값을 몇 개 이상 사용하게 된다면 유지보수하다 실수하기 좋죠.

alex-gibeom replied:
최대한 공통적으로 사용되는 값 들은 따로 뺐어야 됐는데 놓친 것 같습니다..!

클래스 최상단에 static 상수로 선언하였습니다!
```java
private static final Long TEST_ID = 1L;
```

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,118 @@
+
+        int newSize = taskService.getTasks().size();
+
+        // id 값이 정상적으로 1씩 올라가는지 확인
+        assertThat(newId - oldId).isEqualTo(1);
+        
+        // 할일 목록 List 사이즈가 증가했는지 확인
+        assertThat(newSize - oldSize).isEqualTo(1);
+
+        // title이 POSTFIX_TITLE로 잘 들어갔는지 확인
+        assertThat(task.getTitle()).isEqualTo(POSTFIX_TITLE);
```
이런 종류의 테스트를 설명하는 텍스트들을 '구조화'할 방법이 있을지에 대해 생각해 보세요.

alex-gibeom replied:
(7573917) 해당 수정에서 `@Nested`를 이용하여 구조화를 진행해보았습니다!

alex-gibeom replied:
확실히 구조화를 시키니 테스트를 볼 때 가독성이 확 올라가서 좋은 것 같습니다 :)
<img src="https://user-images.githubusercontent.com/59248326/197769531-e88c6515-5907-4559-9900-64c0c3d4d2a0.png" width="50%">

johngrib replied:
이건 따로 알려드리려고 했더니 먼저 조사해서 적용하셨네요!

alex-gibeom replied:
테스트 구조화에 대해 서칭하다가 [리뷰어님의 블로그](https://johngrib.github.io/wiki/junit5-nested/)를 발견해버려서 정독했습니다 :)

johngrib reviewed (commented):
이번주도 화이팅!

alex-gibeom committed:
refactor: TaskService 테스트 구조화

alex-gibeom committed:
refactor: Task 테스트 구조화

alex-gibeom committed:
merge: [Day-2] 유닛 테스트

alex-gibeom committed:
feat: TaskController 웹 테스트 (Mockito) - 상세 조회, 목록 조회 완료

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,200 @@
+        source.setTitle(TEST_TITLE);
+        Task task = taskService.createTask(source);
+
+        assertThat(task.getId()).isEqualTo(TEST_ID);
+        assertThat(task.getTitle()).isEqualTo(TEST_TITLE);
```
여기는 테스트를 준비하는 곳입니다. 검증은 `@Test`에서 하도록 합시다.

alex-gibeom replied:
(26208eb) fixture 데이터 생성 메서드에서 검증 로직 제거하였습니다.

해당 관련 질문은 상단리뷰에 남겨놓았습니다! (https://github.com/CodeSoom/spring-week3-assignment-1/pull/89#discussion_r1004418315)

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,200 @@
+
+    @Nested
+    @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+    class 할_일_목록을_반환할_때 {
+        @Test
+        @DisplayName("할 일 목록을 반환한다")
+        void getTasks() {
+            List<Task> tasks = taskService.getTasks();
+
+            // fixture 데이터가 잘 들어오는지 확인
+            assertThat(tasks).hasSize(1);
```
테스트 설명을 모아놓고 읽으면 좀 이상하게 읽힙니다.

`할 일 목록을 반환할 때, 할 일 목록을 반환한다.`

alex-gibeom replied:
(cfca9ee) 테스트 이름들을 모두 **자연스러운 문장으로 읽히도록** 수정해보았습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,200 @@
+
+        @Nested
+        @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+        class 존재하는_id일_경우 {
```
`존재하는` 보다 `찾을 수 있는`을 사용해 보세요. "존재하는"는 매우 강한 표현입니다.

alex-gibeom replied:
(https://github.com/CodeSoom/spring-week3-assignment-1/commit/cfca9ee412b0913811d71b5a4e60fa387581800a) `존재하는` -> `찾을 수 있는`으로 변경하였습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,200 @@
+        @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+        class 존재하지_않는_id일_경우 {
+            @Test
+            @DisplayName("TaskNotFoundException을 반환한다")
+            void it_returns_taskNotFoundException() {
+                assertThatThrownBy(() -> taskService.getTask(100L))
+                        .isInstanceOf(TaskNotFoundException.class);
```
`getTask`를 호출해도 `TaskNotFoundException`을 반환(return) 하지 않습니다. `getTask`는 `Task`를 리턴하며, 예외적인 상황에서 `TaskNotFoundException`을 던집니다(throw).

alex-gibeom replied:
(https://github.com/CodeSoom/spring-week3-assignment-1/commit/cfca9ee412b0913811d71b5a4e60fa387581800a) `예외를 던진다`로 문구를 수정하였습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,200 @@
+        }
+
+        @Test
+        @DisplayName("생성한 할 일의 id는 직전 값보다 1이 증가한다")
+        void it_returns_created_id() {
+            Task source = taskService.getTask(TEST_ID);
+            Long oldId = source.getId();
+
+            source.setTitle(POSTFIX_TITLE);
+
+            Task task = taskService.createTask(source);
+
+            Long newId = task.getId();
+
+            assertThat(oldId + 1).isEqualTo(newId);
```
이 테스트는 중요하지 않은 것을 테스트하고 있습니다. id 가 1 이 증가하는지 아닌지는 중요한 문제가 아닙니다. unique 한 아이디가 생성되었는지가 중요하다는 점을 고려하고 테스트해보세요.

alex-gibeom replied:
(956a860) 기존 1 증가 테스트 케이스를 삭제한 후, 생성 된 id값이 유니크한지 검증하는 테스트 케이스를 추가하였습니다.

johngrib replied:
좋습니다!

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,200 @@
+            }
+
+            @Test
+            @DisplayName("할 일의 목록 크기는 1 감소한다")
```
이렇게 표현할 수도 있겠지요.

`할 일이 하나 줄어든다.`

alex-gibeom replied:
최대한 압축하여 간추리는 연습도 해야겠습니다.

(cfca9ee) 추천해주신 이름으로 변경하였습니다!

alex-gibeom committed:
refactor: TaskService 테스트 이름을 자연스러운 문장으로 수정

alex-gibeom committed:
feat: TaskServiceTest의 createTask 테스트 케이스 추가 - 생성된 id값이 유니크한지 확인

alex-gibeom committed:
refactor: fixture 데이터 수정 보완

alex-gibeom committed:
feat: TaskController MockMVC 테스트 완료

alex-gibeom committed:
merge: [Day-3] MockMVC 테스트 & 유닛테스트 리팩토링

alex-gibeom commented:
금일 수정 및 개발한 테스트 실행결과를 캡쳐해보았습니다 :)
- 왼쪽: TaskServiceTest
- 오른쪽: TaskControllerWebTest (MockMVC)

<details>
<summary>토글 접기/펼치기</summary>
<div markdown="1">

<img src="https://user-images.githubusercontent.com/59248326/197942931-21e0d590-321c-47e6-936b-a4ead99ede8e.png" width="45%" />  <img src="https://user-images.githubusercontent.com/59248326/198018548-c5383dc1-0b80-4324-99bc-64a8aa9f9fde.png" width="44%" />

</div>
</details>

johngrib commented:
> 금일 수정 및 개발한 테스트 실행결과를 캡쳐해보았습니다 :)
> 
> * 왼쪽: TaskServiceTest
> * 오른쪽: TaskControllerWebTest (MockMVC)
> 

이렇게 작업하고 보면 뿌듯하고 기분도 좋죠 ㅎㅎ

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+        taskService.createTask(taskSourceWithSetTitle);
+    }
+
+    @Nested
+    @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+    class getTasks_메서드는 {
+        @Test
+        @DisplayName("할 일 목록을 반환한다")
+        void getTasks() {
+            List<Task> tasks = taskService.getTasks();
+
+            // fixture 데이터가 잘 들어오는지 확인
+            assertThat(tasks).hasSize(1);
+        }
+    }
```
할 일 목록이 1개를 초과하는 경우에 대한 테스트도 추가해 봅시다.

alex-gibeom replied:
`@ParameterizedTest`와 `@ValueSource`를 통해 테스트 반복 기능을 사용하여,
할 일 목록 1개 초과인 경우의 테스트 케이스를 추가해보았습니다.

**[ getTasks() 테스트 수정 결과 ]**

![image](https://user-images.githubusercontent.com/59248326/198208764-5d437cd7-a71a-4e1a-a6f1-4a0c81835fb9.png)


#### [ 질문 ]
`@ParameterizedTest`를 실행하는 테스트 메서드의 이름을 `할 일이 n개라면`으로 작성해보았는데, 좀 더 네이밍이 있을 지 궁금합니다.
테스트 반복을 실행하는 메서드 이름이 참 애매하네요 😂

<br>

johngrib replied:
이런 경우에는 `n개의 할 일 목록을 리턴한다`로 합쳐주는 것이 좋겠습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+
+    @Nested
+    @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+    class createTask_메서드는 {
+
+        @Nested
+        @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+        class 할_일을_생성할_경우 {
+            @Test
+            @DisplayName("생성한 할 일을 반환한다")
```
두 문장을 비교해 봅시다. 두번째와 세번째 문장은 제가 작성한 것입니다.

- `createTask` 메서드는, 할 일을 생성할 경우, 생성한 할 일을 반환한다
- `createTask` 메서드는, 새로운 할 일이 주어지면, 할 일을 생성하고 리턴합니다.
- `createTask` 메서드는, 할 일을 생성하고 리턴합니다.

Describe 절에서는 테스트 대상이 무엇인지를 명시하고, Context 절에서는 테스트 대상에 무엇을 주거나 어떤 환경을 제공하는지를 명시해 준다고 생각해 봅시다. createTask 메서드는 할 일을 생성하지 않는 경우를 생각하기 어렵죠. 따라서 크게 의미있는 컨텍스트로 보이지 않습니다. 하지만 다음과 같은 테스트를 구성한다면 어떨까요?

- createTask 메서드는
  - 새로운 할 일이 주어지면
    - 할 일을 저장하고 리턴한다
  - null이 주어지면
    - 예외를 던진다
  - 비속어/금지어가 포함된 할 일이 주어지면
    - 할 일을 저장하고 경고 카운트를 증가시킨 다음 저장한 할 일을 리턴한다
  - 중복된 내용을 갖는 할 일이 주어지면
    - '할 일 중복'을 의미하는 예외를 던진다
 
이 리스트가 일종의 예시라는 것을 감안하고 읽어보셔야 합니다. 왜냐하면 지금 코드 베이스 기준에서는 세번째/네번째 컨텍스트가 의미가 없기 때문이죠. 하지만 이 스타일의 테스트 코드가 개발/운영/유지보수를 거치며 컨텍스트가 증가하는 상황에 대응하기 좋다는 것은 짐작할 수 있을 것입니다.

alex-gibeom replied:
Context 절이 `무엇을 주거나`, `어떤 환경을 제공`하는지의 관점으로 봐야하는군요.

상황에 따라 다르겠지만 파라미터가 없다면 `~ 때`, 파라미터가 있다면 `~ 주어지면`으로 Context절을 시작해주면,
하위 추가 케이스에도 유연하게 대응할 수 있을 것 같네요 🙂

<br>

(508e457) 추천해주신 스타일로 Context 절 문구를 수정, 보완해보았습니다.

<img src="https://user-images.githubusercontent.com/59248326/198231963-8106774e-4a45-47f9-8573-9b1bed9d6136.png" width="40%" />

johngrib replied:
굉징히 읽기 좋습니다. 잘 하셨습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+                assertThat(task.getTitle()).isEqualTo(TEST_TITLE);
+            }
+
+            @Test
+            @DisplayName("할 일이 하나 늘어난다")
+            void it_returns_tasks_size() {
+                int oldSize = taskService.getTasks().size();
+
+                taskService.createTask(taskSourceWithSetTitle);
+
+                int newSize = taskService.getTasks().size();
+
+                assertThat(newSize - oldSize).isEqualTo(1);
```
잘 만든 테스트입니다.

alex-gibeom replied:
감사합니다 😊

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+                assertThat(task.getTitle()).isEqualTo(TEST_TITLE + POSTFIX_TITLE);
+            }
+
+            @Test
+            @DisplayName("할 일의 id값은 수정하지 않는다")
+            void it_returns_id() {
+                taskSourceWithSetTitle.setId(TEST_NOT_EXSIT_ID);
+
+                Task task = taskService.updateTask(TEST_ID, taskSourceWithSetTitle);
+
+                assertThat(task.getId()).isEqualTo(TEST_ID);
+            }
```
상식을 토대로 판단해서 이런 테스트를 지우는 선택도 가능합니다. 이 테스트를 삭제하라는 뜻은 아닙니다. 삭제해도 납득하는 방향으로 팀과 커뮤니케이션 할 수 있다는 뜻입니다. 자, 이 테스트를 어떻게 하시겠습니까?

alex-gibeom replied:
보통 데이터 수정을 할 때, pk값은 수정을 허용하지 않습니다.
pk 값 수정 자체는 가능하나 아래와 같이 많은 고려 대상이 필요하므로, 원칙적으로 수정을 허용하지 않습니다.
- 수정할 값이 유니크한 값 인지?
- 그 키를 참조하고 있는 자식 테이블은 존재하는지? (외래키) 등

이런 이유들로 인해 일반적으로 데이터 수정 API 요청 body를 위한 `RequestDto`에는,
pk 필드를 포함하지 않고 `@PathVariable`을 통해 pk 값을 전달받습니다.
따라서 **Entity로 변환**될 때 pk 값은 포함되지 않으므로, 기존 데이터가 수정될 때 **pk 값은 반영이 되지 않습니다**.

<br>

#### [실제 메서드 내부 동작]
해당 service의 `updateTask()` 메서드의 내부 동작은 첫번째 인자인 `TEST_ID`를 통해 기존 데이터를 가져온 후,
두번째 인자인 `taskSourceWithSetTitle`를 통해 title만 변경하도록 개발되어 있습니다.
```java
public Task updateTask(Long id, Task source) {
        Task task = getTask(id);
        task.setTitle(source.getTitle());   // title만 변경

        return task;
    }
```

#### [결론]
따라서 id가 수정되는 상황은  **일어나지 않을 테스트 케이스**이므로 삭제해도 되는 테스트인 것 같습니다.

alex-gibeom replied:
위와 같은 이유로 `@Disabled("")` 기능을 통해서 삭제하지 않고, 테스트를 비활성화 시켰습니다.

```java
@Test
@DisplayName("할 일의 id값은 수정하지 않는다")
@Disabled("일어나지 않을 상황을 테스트합니다")
void it_returns_id() {
    taskSource.setId(TEST_NOT_EXSIT_ID);

    Task task = taskService.updateTask(TEST_ID, taskSource);

    assertThat(task.getId()).isEqualTo(TEST_ID);
}
```

johngrib replied:
좋습니다. 의사결정과정을 글로 작성하셨는데요, 이런 식으로 작성하는 것도 실력이어서 자꾸 반복하시면 좋은 연습이 될 것입니다.

alex-gibeom replied:
계속해서 연습해봐야겠습니다.
감사합니다 😊

johngrib reviewed (commented):
굉장히 빨리 적응하시는군요!

alex-gibeom committed:
refactor: TaskServiceTest 클래스 내부의 할 일 생성용 인스턴스명 수정 (taskSourceWithSetTitle -> taskSource)

alex-gibeom committed:
feat: TaskService getTasks() 테스트 케이스 추가 (할 일이 1개 이상인 경우)

alex-gibeom committed:
refactor: 테스트 문구(Context 절) 수정

alex-gibeom committed:
feat: Task 클래스 테스트 케이스 추가

alex-gibeom committed:
merge:  [Day-4] 테스트 케이스 추가 및 Context 문구 수정, 보완

alex-gibeom commented:
금일 수정, 보완한 테스트 명을 캡처해보았습니다.

#### 좌측: TaskController 클래스      &nbsp;&nbsp;&nbsp;     -------       &nbsp;&nbsp;&nbsp;        우측: Task 클래스

<details>
<summary>토글 접기/펼치기</summary>
<div markdown="1">

<img src="https://user-images.githubusercontent.com/59248326/198233213-4e878ae9-1f5b-489d-b59d-56e59b964fcc.png" height="40%">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <img src="https://user-images.githubusercontent.com/59248326/198233364-9196f48d-0d50-49cf-a05f-89eb99221880.png" height="55%">

</div>
</details>

<br>

#### TaskService 클래스

<details>
<summary>토글 접기/펼치기</summary>
<div markdown="1">

<img src="https://user-images.githubusercontent.com/59248326/198233122-cd18f300-908c-4c0e-a1fc-698c6aa6e945.png" height="40%"> 

</div>
</details>

johngrib commented:
<img src="https://user-images.githubusercontent.com/59248326/198233364-9196f48d-0d50-49cf-a05f-89eb99221880.png" height="55%">

`hashCode` 메서드에 대한 테스트 설명이 뭔가 이상한데요? `hashCode`가 `boolean`을 리턴하나요?

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,263 @@
+
+    /**
+     * 각 테스트에서 필요한 fixture 데이터를 생성합니다.
+     *  - 할 일 1개 생성
+     *  - title만 세팅되어 있는 taskSource 변수 선언
```
지금은 주석이 메소드의 구현에 의존하고 있습니다. 이 메소드의 내용이 바뀌게 되었을 때 작업한 사람이 주석의 내용을 수정하지 않았다고 생각해 봅시다.  
```suggestion
```

alex-gibeom replied:
메서드 변경을 대비해서 주석은 추상적으로 작성해야겠군요.
주석이 구현에 의존되지 않게 조심해야겠습니다.

감사합니다 :)

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,263 @@
+        @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+        class 할_일이_있을_때 {
+            @DisplayName("할 일이 n개라면")
+            @ParameterizedTest(name = "{arguments}개의 할 일 목록을 리턴한다")
+            @ValueSource(ints = {1, 77, 1027})
+            void it_returns_tasks(int createCount) {
+
+                createTaskUntilCount(createCount);
+
+                List<Task> tasks = taskService.getTasks();
+
+                assertThat(tasks).hasSize(createCount);
```
설명 String을 `assert`에서 명시하는 방법이 있을 겁니다. 조사해 보세요.

alex-gibeom replied:
많이 찾아보고 구현해보고 있는데, 혹시 추천해주신 기능이 JUnit 5.8부터 지원하는 Named API를 의도하신게 맞는지 궁금합니다!  [(참고중인 자료)](https://jschmitz.dev/posts/how_to_improve_junit_displaynames_with_the_new_named_api/)

johngrib replied:
아뇨. 제가 말하는 건 이런 겁니다.

https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/Assertions.html

![image](https://user-images.githubusercontent.com/1855714/198586473-549c3423-1494-4455-9315-867796788992.png)

alex-gibeom replied:
아 테스트가 실패할 경우 실패 메시지를 제공하는 Junit 5기능을 말씀하시는거였군요.

서칭 해본 결과, 해당 내용은 아래와 같이 작성할 수 있을 것 같습니다.
```java
// Assertions
assertEquals(createCount, tasks.size(),
                        String.format("%d개의 할 일을 리턴해야합니다", createCount));

// AssertJ (코드 일관성을 위해 해당 구문으로 커밋하였습니다)
assertThat(tasks)
        .as("%d개의 할 일 목록을 리턴한다", createCount)  // description (@ParameterizedTest에서 테스트 설명을 대신하여, 실제 커밋에서는 삭제하였습니다)
        .withFailMessage("%d개의 할 일을 리턴해야합니다", createCount)  // newErrorMessage
        .hasSize(createCount);
```

**[실제 커밋한 코드]** (9ca8798)
```java
assertThat(tasks)
        .withFailMessage("%d개의 할 일을 리턴해야합니다", createCount)
        .hasSize(createCount);
```

추가로 `Supplier<String>`을 통해 실패 메시지의 비용을 줄일 수 있다고 합니다! [(Lazy error message overriding)](https://assertj.github.io/doc/#assertj-core-lazy-error-message)

<br>

제가 **오해하고 있었던 사항**은 아래의 문구들을 없앨 수 있게, 메서드 구현부에서 **테스트 설명 string을 설정**해보라는 뜻인 줄 알았습니다.

<img src="https://user-images.githubusercontent.com/59248326/198803596-bf119f50-0072-4e7d-b226-ee0b1f3ce550.png" width="30%" />

따라서 해당 내용을 해보려고 시도했는데, 제가 사용했던 `@ParameterizedTest`에서는 어디에서 설명 string을 설정하든 name에 들어갈 문구에 **공백은 허용하지 않는**다고 해서 실패했었습니다.

> The display name to be used for individual invocations of the parameterized test; never blank or consisting solely of whitespace. [(공식문서)](https://junit.org/junit5/docs/5.2.0/api/org/junit/jupiter/params/ParameterizedTest.html)

johngrib replied:
네. `@ParameterizedTest`를 사용하는 경우엔 어쩔 수 없죠.

johngrib commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,174 @@
+        task = new Task();
+    }
+
+    @Nested
+    @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+    class Getter_메서드는 {
+
+        @Nested
+        @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+        class 데이터가_없을_때 {
+            @Test
+            @DisplayName("null을 리턴한다")
+            void it_returns_null() {
+                assertThat(task.getId())
+                        .isNull();
+
+                assertThat(task.getTitle())
+                        .isNull();
+            }
+        }
```
뻔하고 단순한 getter 구현이라면 굳이 테스트하지 않아도 됩니다.

alex-gibeom replied:
미션이 테스트 커버리지 100%라고 해서 작성해보았는데, Getter 테스트를 제거해도 Setter 테스트에서 체크하기 때문에 커버리지로 인정되나보네요.

(39e988f) `@Disabled` 처리 하였습니다.

johngrib replied:
미션에 커버리지 100%가 있긴 하지만 커버리지 정책이 회사마다 다 다르기 때문에 다양하게 고민해보시면 좋습니다.

alex-gibeom replied:
안그래도 최근에 백기선님 유튜브 라이브에서 어떤 사연자가 **1주일에 신규 기능 하나 배포**하는데, **테스트 커버리지가 80프로 이하이면 merge가 안되는 설정**이 있다고 해서 "**과연 커버리지가 제대로 된 검증을 증명할 수 있는 것일까?**” 라는 생각을 해봤었습니다. (배포 주기가 사기급이라...)

리뷰어님 덕분에 오늘 커비리지에 대해서 좀 알아봤는데, 커버리지 **수치에만 매달릴 경우** 오히려 **코드 품질이 저하**될 수 있을 것 같다는 생각이 듭니다. (공장처럼 무가치한 테스트 케이스 작성 우려) - [(공감됐던 글)](http://aeternum.egloos.com/1209470)

커버리지 종류가 **라인**, **결정**, **조건**, **패스** 등 많은 종류가 있네요.
대표적으로 많이 사용된다는 라인 커버리지는 모든 경우를 커버해주지 못해서,
안전한 테스트 기준이 못될 것 같은 생각이 드는데 회사에서의 테스트 경험이 없어서 감이 안잡히네요 😂

리뷰어님은 커버리지 정책을 어떤식으로 사용하시는지 여쭤봐도 될까요?

johngrib replied:
저는 기계적인 커버리지는 중요하지 않다고 생각해서 신경쓰지 않습니다. 핵심 기능들을 다 테스트하고 있는지와, 운영 중에 발견된 버그에 대해 테스트 코드를 추가해 "과거에 해결한 버그가 되살아나면 즉시 발견할 수 있도록" 하는 것을 중요하게 생각합니다. 이렇게 버그에 대한 테스트 코드를 쌓아간다면 경험상 기계적으로 테스트 커버리지를 챙기는 것보다 더 합리적으로 운영할 수 있었어요.

alex-gibeom replied:
그렇군요.
제가 생각해도 무작정 커버리지 수치를 강제하기보다는, 핵심적인 중요한 기능들은 다 테스트하고 운영중에 발생한 버그들을 위주로 테스트 코드를 쌓아가는 방법이 실무에서는 효율적인 방법일 것 같네요.
답변 감사드립니다 :)

johngrib reviewed (commented):
너무 열심히 잘 작업하셔서 리뷰 남길 게 별로 없을 정도로군요 ㅎㅎ
추가 과제를 드리겠습니다. 중복 할 일이 발생할 경우, "이미 등록된 할 일입니다"라고 응답하는 기능을 추가해 봅시다.

alex-gibeom committed:
refactor: 테스트 문구(Context 절) 수정 및 fixture 주석 수정

alex-gibeom commented:
> `hashCode` 메서드에 대한 테스트 설명이 뭔가 이상한데요? `hashCode`가 `boolean`을 리턴하나요?

<br>

앗 테스트를 하루에 여러 개 짜다 보면 마지막쯤 테스트는 디테일이 좋지 않네요.. 😂
집중력이 떨어지나 봅니다.

(cdec005) 수정하였습니다 감사합니다 :)

![image](https://user-images.githubusercontent.com/59248326/198550572-fffad5cd-0c98-4efc-a221-ca5ac1847c74.png)

alex-gibeom committed:
fix: Task Hashcode 메서드 테스트 문구 수정

alex-gibeom committed:
refactor: Task getter 메서드 테스트 Disabled 처리

alex-gibeom committed:
feat: 공통 Exception 구축

alex-gibeom committed:
feat: 중복된 할 일 등록 제한 기능 추가

alex-gibeom committed:
merge: [Day-5] 테스트 케이스 추가 및 리팩토링

alex-gibeom commented:
> 너무 열심히 잘 작업하셔서 리뷰 남길 게 별로 없을 정도로군요 ㅎㅎ 추가 과제를 드리겠습니다. 중복 할 일이 발생할 경우, "이미 등록된 할 일입니다"라고 응답하는 기능을 추가해 봅시다.

(6a87384), (8204581) 감사하게 주신 추가 과제를 구현해 보았습니다.

<img src="https://user-images.githubusercontent.com/59248326/198581280-01cae79f-2e51-417a-a177-72fee1634561.png" width="50%">

<br>

추가로 금일 테스트 수정, 보완한 내용 캡처해보았습니다.
TaskController는 변동 사항 없으므로 제외하였습니다.

<details>
<summary>토글 접기/펼치기</summary>
<div markdown="1">

<img src="https://user-images.githubusercontent.com/59248326/198581673-3744d5f6-0a5e-40f1-91b8-cf294f2a3dc8.png" width="50%"> <img src="https://user-images.githubusercontent.com/59248326/198581780-6a97e85a-9cb3-45b7-a17b-913cf4574ba4.png" width="40%">


</div>
</details>

johngrib commented on app/src/main/java/com/codesoom/assignment/exceptions/TaskAlreadyExistException.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * 중복된 할 일을 등록할 때 던집니다.
+ */
+public class TaskAlreadyExistException extends CommonException {
```
이 정도가 적절할 것 같습니다. 어쩌면 Task가 이미 존재하는 것은 문제가 아니게 될 지도 모릅니다.
```suggestion
public class TaskDuplicationException extends CommonException {
```

alex-gibeom replied:
(47eecc5) 수정하였습니다.

johngrib reviewed (approved):
LGTM

johngrib commented:
여기에서 이번주를 마무리하기는 좀 아쉬우니 추가 과제를 더 드리겠습니다. 

지금은 Task가 Task로만 있고 언제까지 할 일인지, 얼마나 높은 중요도를 갖고 있는지 등이 드러나 있지 않습니다. 이런 기능들에 대해 자유롭게 생각해 보시고 추가해 보세요.

alex-gibeom committed:
refactor: 할 일 목록 반환 메서드에서 반복 테스트에 실패 메시지 추가

alex-gibeom committed:
refactor: 중복 할 일 예외명 변경 (TaskAlreadyExist -> TaskDuplication)

giibeom commented:
안녕하세요 리뷰어님 :)
덕분에 이번주도 많은 것을 배우고 있습니다 😊

한가지 궁금한 점이 있는데요.
제가 이번 주 미션 중 Controller 테스트를 두개 만들어야된다는것을 좀 늦게 인지했습니다.
- TaskController 유닛 테스트
- TaskController MockMVC 테스트

현재는 TaskController를 MockMVC 테스트만 짜서 유닛 테스트를 작성하려 했는데,
문득 드는 생각이 TaskService 테스트와 별다를게 없다는 생각이 들었습니다.
현재 컨트롤러에 작성된 모든 코드가 동일하게 **호출한 service의 반환값을 리턴**하는 것 말고는 별다른 동작이 없는데요.
이런 경우에도 Mock 테스트와, Mock을 안쓰는 유닛 테스트도 같이 작성하는 것이 좋은건지 궁금합니다. (두 버전의 테스트를 다 작성해야하는지)

개인적인 견해는 Controller의 역할은 보통 요청과 응답에 대한 책임만 담당하므로 관련 Service를 호출 후 응답 값을 반환하는 코드만 있는게 대부분인 것 같아서,
Mock 테스트만 있어도 될 것 같은데 리뷰어님의 의견이 궁금합니다!

alex-gibeom committed:
feat: TaskController 유닛 테스트 추가

alex-gibeom committed:
merge: [Day-6] TaskController 유닛 테스트 추가 및 리팩토링

alex-gibeom commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,224 @@
+import static org.mockito.Mockito.verify;
+
+@DisplayName("TaskController 유닛 테스트")
+class TaskControllerTest {
```
안녕하세요 리뷰어님 :) 덕분에 이번주도 많은 것을 배우고 있습니다 😊

한가지 궁금한 점이 있는데요.제가 이번 주 미션 중 Controller 테스트를 두개 만들어야된다는것을 좀 늦게 인지했습니다.
- TaskController 유닛 테스트
- TaskController MockMVC 테스트

원래는 TaskController를 MockMVC 테스트만 짜서, 오늘 급하게 유닛 테스트를 작성해보았는데, 문득 드는 생각이 TaskService 테스트와 별다를게 없다는 생각이 들었습니다. 
현재 컨트롤러에 작성된 모든 코드가 동일하게 **호출한 service의 반환값을 리턴**하는 것 말고는 별다른 동작이 없는데요.
따라서 TaskServiceTest에 비교하여 추가 된 코드는 각 테스트마다 `verify(taskService).getTasks();` 같이 `verify` 코드만 추가하였습니다.

보통 Mock 테스트와, Mock을 안쓰는 유닛 테스트 2개 다 작성하는 것이 좋은건지 궁금합니다. (두 버전의 테스트를 다 작성해야하는지)

johngrib replied:
이것 또한 조직의 개발 철학/컨벤션에 따라 달라지는 경우가 많습니다. 저는 개인적으로 Mock 테스트를 싫어해서 가능한 한 안 쓰면서(정말 필요한 곳에만 쓰는 방식) 작업하는 것이 바람직하다고 생각합니다. 그러나 싫어함에도 불구하고 반드시 써야 하는 경우도 있다고 생각하며, 실제로 회사에서 일할 때에도 의도적으로 mock 테스트를 작성하는 일도 흔합니다. 이 주제에 대해서는 다른 분의 PR에서 제가 나눈 대화가 있으니 한번쯤 읽어보시는 것도 좋을 것 같군요.

https://github.com/CodeSoom/spring-week3-assignment-1/pull/58#discussion_r845062254

한편 이에 대해 함께 읽어볼만한 괜찮은 책도 있습니다. "구글 엔지니어는 이렇게 일한다"의 12장과 13장을 읽어보세요. 저는 이 책을 읽을 때 제 생각과 비슷한 면이 꽤 있어서 기분 좋게 읽은 편이긴 하지만, 항상 정답은 없다는 것을 염두에 두시고 가급적이면 균형적인 시각을 가지려 노력하며 읽어보세요.

alex-gibeom replied:
결국 TDD를 할 때 미구현 모듈 사용 시, 외부 모듈 사용 시 등의 상황에서는 given을 적절히 사용해야되지만, 최소한 테스트를 학습할 때는 최대한 given을 지양하면서 사용해봐야겠군요.
테스트를 많이 짜보면서 given의 적절한 필요 시기를 판단하는 능력도 길러봐야겠네요 😂

추가로 `@Mock`과 별개로 `@Spy`는 본래 실제 객체를 사용하면서 필요한 부분만 given으로 사용할 수도 있군요.
`@Spy`를 사용하면서 적절한 시기에 given을 사용하는 연습을 해봐야겠습니다.

 "구글 엔지니어는 이렇게 일한다" 책도 구매해봐야겠습니다.
감사합니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,224 @@
+        @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+        class 할_일이_있을_때 {
+            @Test
+            @DisplayName("비어있지 않은 ArrayList를 리턴한다")
```
`비어있지 않은 리스트를 리턴한다` 정도도 적절하겠죠. 리턴 타입이 정확히 ArrayList인지는 이 문제에서 별로 중요한 문제가 아닐 수 있습니다.

alex-gibeom replied:
그렇겠네요. 집중력을 끝까지 잃지 않는 연습을 해야겠습니다.
(0442952) 수정하였습니다.

alex-gibeom committed:
feat: TaskController 유닛 테스트 mock 제거 (spy로 변경)

alex-gibeom committed:
refactor: TaskController MockMVC 테스트 내부 given절 `@BeforeEach` 메서드에 응집

alex-gibeom committed:
reafactor: Task 객체 setter 제거 작업 (불변 객체를 위한 빌더 사용)

alex-gibeom committed:
merge:  [Day-7] TaskController 유닛 테스트 mock 지양하기

alex-gibeom commented:
> 여기에서 이번주를 마무리하기는 좀 아쉬우니 추가 과제를 더 드리겠습니다.
> 
> 지금은 Task가 Task로만 있고 언제까지 할 일인지, 얼마나 높은 중요도를 갖고 있는지 등이 드러나 있지 않습니다. 이런 기능들에 대해 자유롭게 생각해 보시고 추가해 보세요.

할 일의 종료 시간을 추가하는 기능을 만들어 보아서, TaskService의 createTask(), updateTask(), getTasks() 3개에 대한 테스트가 끝났는데 TaskControllerWebTest에서 테스트가 깨지네요...
결국 시간 부족으로 리뷰 받을 때 까지는 완성하지 못했습니다 😂

종료 시간 변수인 endTime을 LocalDateTime 타입으로 지정했는데 `mockMvc.perform()` 에서 `.content()` 에서 깨지네요.
ObjectMapper를 통해 변환 후 `content()`에 값을 집어넣어주는데 그 과정에서 400 에러가 발생했습니다
오늘 해당 기능까진 마무리 해보겠습니다!

원래 구성되어있던 코드에서 추가 기능을 개발하면서 Test 코드를 고치다보니 느낀 것이,
각 테스트는 오히려 독립적으로 분리해놓는 것이 좋을 것 같다는 생각도 들긴 하네요 😂
근데 또 한편으로는 말씀하신 것처럼 mock을 사용할 경우, 기능을 추가할 시 불이 나도 울리지 않는 사이렌 처럼 테스트로 에러를 못잡는 상황도 있을 수 있을 것 같습니다.

johngrib reviewed (commented):
어느새 일요일이 됐네요. 추가 과제도 성실하게 해 주셔서 정말 보람찬 한 주가 된 것 같습니다. 주간 회고를 작성해서 공유해 주세요.

alex-gibeom committed:
feat: 할 일의 종료시간 기능 추가 (목록 조회, 생성, 수정)

- 종료 시간을 따로 입력 안한 경우 기본 값 LocalDateTime.Max() (999999999-12-31........)
- 종료 임박 순으로 목록 조회 (종료 시간이 같을 경우 id로 정렬)
- 생성, 수정 시 종료 시간 반영

alex-gibeom commented:
덕분에 한주동안 많은 것을 배운 것 같습니다 :)
감사합니다 🙇‍♂️🙇‍♂️

johngrib commented on app/src/main/java/com/codesoom/assignment/application/TaskService.java:
```diff
@@ -14,6 +14,15 @@ public class TaskService {
 
     public List<Task> getTasks() {
+
+        tasks.sort((o1, o2) -> {
+            if (o1.getEndTime().equals(o2.getEndTime())) {
+                return o1.getId().compareTo(o2.getId());
+            }
+
+            return o1.getEndTime().compareTo(o2.getEndTime());
+        });
+
```
Task 클래스가 Comparable 인터페이스를 구현하게 하는 방법도 고려해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -3,18 +3,25 @@
 
     @Builder
-    public Task(Long id, String title) {
+    public Task(Long id, String title, LocalDateTime endTime) {
```
만약 이 할 일 서비스가 국제적인 서비스가 된다면 어떤 문제가 있을까요?
